src/CodeGenerator/AbstractInjector.php

```diff
@@ -49,14 +52,23 @@ public function __construct(InjectorInterface $injector, ContainerInterface $con
      */
     abstract protected function loadFactoryList() : void;
 
+    private function setFactory(string $type, FactoryInterface $factory) : void
+    {
+        $this->factoryInstances[$type] = $factory;
+    }
+
     private function getFactory($type) : FactoryInterface
     {
-        if (is_string($this->factories[$type])) {
-            $factory = $this->factories[$type];
-            $this->factories[$type] = new $factory();
+        if (isset($this->factoryInstances[$type])) {
+            return $this->factoryInstances[$type];
         }
 
-        return $this->factories[$type];
+        $factoryClass = $this->factories[$type];
+        $factory = ($factoryClass instanceof FactoryInterface) ? $factoryClass : new $factoryClass();
```

Comment:
Unfortunately we still need this check for BC.

Reply:
Yeah, this change is a BC break indeed, sorry for bringing you to implement it this way.

I think we need to revert to single array of `string[]|object[]` for now

Reply:
I guess my comment caused more confusion than clarifying things here, sorry 'bout that.

I just wanted to point out why this part looks so odd: because it actually keeps BC instead of breaking it. It assures that the `$factories` array can contain `string[]|FacotryInterface[]` exactly the same way as before this PR (tests failed when I changed `$factories` to a pure string array). 

Your hint for two separate arrays was perfectly fine though.